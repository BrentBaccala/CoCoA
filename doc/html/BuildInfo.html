<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.sf.net">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      BuildInfo</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<P ALIGN="center"><CENTER><H1>      BuildInfo</H1>
<FONT SIZE="4">
<I>      &copy;  2005,2007 John Abbott</I><BR>
      GNU Free Documentation License, Version 1.2
</FONT></CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
    <UL>
    <LI><A HREF="#toc1">Examples</A>
    <LI><A HREF="#toc2">User documentation</A>
    <LI><A HREF="#toc3">Maintainer documentation</A>
    <LI><A HREF="#toc4">Bugs, Shortcomings and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>
<A NAME="toc1"></A>
<H2>Examples</H2>
<UL>
<LI><A HREF="../../examples/index.html#ex-BuildInfo.C">ex-BuildInfo.C</A>
<LI><A HREF="../../examples/index.html#ex-limits.C">ex-limits.C</A>
</UL>

<P></P>
<A NAME="toc2"></A>
<H2>User documentation</H2>
<P>
The constant in <CODE>BuildInfo</CODE> allows you to find out which version of
CoCoALib you are using.  The function <CODE>BuildInfo::PrintAll</CODE> prints out all
the build information on the <CODE>ostream</CODE> passed in as argument -- you should
include this information whenever you report a bug.
</P>
<P>
There is one string constant which contains the version number of the
library: it is of the form <CODE>A.bcde</CODE> where <CODE>A</CODE> is the major
version, <CODE>bc</CODE> is the minor version, and <CODE>de</CODE> is the patch level.
Note that there are always precisely 4 digits after the point (even if
they are all zero).
</P>
<UL>
<LI><CODE>BuildInfo::version</CODE> -- a C string containing the CoCoALib version number.
</UL>

<P></P>
<P>
NOTE: if you happen upon a copy of libcocoa.a and want to find out which
version it is, you can use the following Unix/Linux command:
</P>
<PRE>
    strings libcocoa.a | egrep "CoCoA::BuildInfo"
</PRE>
<P></P>
<P>
This should print out three lines informing you of the library version,
the compiler used, and the compiler flags used when creating libcocoa.a.
</P>
<A NAME="toc3"></A>
<H2>Maintainer documentation</H2>
<P>
I chose to put the constants and function in their own namespace to
emphasise that they go together.
</P>
<P>
There are actually four string constants, but only one is supposed to
be publicly accessible (because I cannot imagine why anyone would want
access to the other three).  I made the constants C strings because it
seemed simpler than using C++ strings.  The three constants
<CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and <CODE>CompilerFlagsMesg</CODE> contain
the substring <CODE>CoCoA::BuildInfo</CODE> so that the egrep trick described
above will produce the version/compiler information directly.
</P>
<P>
I made <CODE>BuildInfo::PrintAll</CODE> leave a blank line before and after its
message so that it would stand out better from other output produced
by the program.
</P>
<A NAME="toc4"></A>
<H2>Bugs, Shortcomings and other ideas</H2>
<P>
The constants are not C++ strings -- is this really a bug?
</P>
<P>
Should the three constants <CODE>VersionMesg</CODE>, <CODE>CompilerMesg</CODE>, and
[CompilerFlagsMesg] be hidden or public?  Until someone convinces me there
is a good reason to make them public, they'll stay private.
</P>

<!-- html code generated by txt2tags 2.5 (http://txt2tags.sf.net) -->
<!-- cmdline: txt2tags -t html -o html/BuildInfo.html -i txt/BuildInfo.txt -->
</BODY></HTML>
