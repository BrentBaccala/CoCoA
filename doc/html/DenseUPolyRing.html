<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">
<HTML>
<HEAD>
<META NAME="generator" CONTENT="http://txt2tags.org">
<LINK REL="stylesheet" TYPE="text/css" HREF="cocoalib-doc.css">
<TITLE>      DenseUPolyRing</TITLE>
</HEAD><BODY BGCOLOR="white" TEXT="black">
<CENTER>
<H1>      DenseUPolyRing</H1>
<FONT SIZE="4"><I>      &copy;  2007  Anna Bigatti</I></FONT><BR>
<FONT SIZE="4">      GNU Free Documentation License, Version 1.2</FONT>
</CENTER>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>

    <UL>
    <LI><A HREF="#toc1">User documentation for DenseUPolyRing</A>
      <UL>
      <LI><A HREF="#constructors">Pseudo-constructors</A>
      <LI><A HREF="#query">Query and cast</A>
      <LI><A HREF="#operations">Operations on a DenseUPolyRing</A>
      </UL>
    <LI><A HREF="#toc5">Maintainer documentation for DenseUPolyRing</A>
    <LI><A HREF="#toc6">Bugs, Shortcomings and other ideas</A>
    </UL>

<P></P>
<HR NOSHADE SIZE=1>
<P></P>
<P>
      <center><a href="index.html">CoCoALib Documentation Index</a></center>
</P>

<A NAME="toc1"></A>
<H2>User documentation for DenseUPolyRing</H2>

<P>
<CODE>DenseUPolyRing</CODE> is an abstract class (inheriting from <A HREF="PolyRing.html"><CODE>PolyRing</CODE></A>)
representing rings of univariate polynomials written with
*dense representation*:
a polynomial is viewed abstractly as a vector of coefficients
(belonging to the coefficient ring).  Indices are <CODE>long</CODE>.
All meaningful coefficients are in the positions 0 to <CODE>deg</CODE>.
Between deg+1 and size-1 the coefficients are guaranteed to be valid
and valued 0.
</P>
<P>
Zero polynomial is represented with <CODE>myDegPlus1 = 0</CODE>.
</P>
<P>
See <A HREF="RingElem.html#DenseUPolyRing"><CODE>RingElem</CODE> DenseUPolyRing</A> for operations on its elements.
</P>

<A NAME="constructors"></A>
<H3>Pseudo-constructors</H3>

<UL>
<LI><CODE>NewPolyRing_DUP(CoeffRing)</CODE> -- default indet name is <CODE>x</CODE>
<LI><CODE>NewPolyRing_DUP(CoeffRing, IndetName)</CODE>
<LI><CODE>NewPolyRing_DUP(CoeffRing, IndetName, MinCapacity)</CODE> --
         allows to specify the minimum capacity for the coefficient
         vector (in order to minimize reallocations)
<LI><CODE>DenseUPolyRing(R)</CODE> -- sort of downcast the ring <CODE>R</CODE> to a dense univariate poly ring;
will throw an exception of type <CODE>CoCoAError</CODE> with code <CODE>ERR::NotDenseUPolyRing</CODE> if needed.
</UL>

<A NAME="query"></A>
<H3>Query and cast</H3>

<P>
Let <CODE>R</CODE> be an object of type <A HREF="ring.html"><CODE>ring</CODE></A>.
</P>

<UL>
<LI><CODE>IsDenseUPolyRing(R)</CODE> -- <CODE>true</CODE> if <CODE>R</CODE> is actually <CODE>DenseUPolyRing</CODE>
<LI><CODE>DenseUPolyRingPtr(R)</CODE> -- get pointer to a <CODE>DenseUPolyRing</CODE> (for calling mem fns);
will throw an exception of type <CODE>CoCoAError</CODE> with code <CODE>ERR::NotDenseUPolyRing</CODE> if needed.
</UL>

<A NAME="operations"></A>
<H3>Operations on a DenseUPolyRing</H3>

<P>
See <A HREF="PolyRing.html#operations"><CODE>PolyRing</CODE> operations</A>.
</P>

<A NAME="toc5"></A>
<H2>Maintainer documentation for DenseUPolyRing</H2>

<A NAME="toc6"></A>
<H2>Bugs, Shortcomings and other ideas</H2>

<P>
No maintainer doc!
</P>

<!-- html code generated by txt2tags 2.6 (http://txt2tags.org) -->
<!-- cmdline: txt2tags -t html -o html/DenseUPolyRing.html -i txt/DenseUPolyRing.txt -->
</BODY></HTML>
