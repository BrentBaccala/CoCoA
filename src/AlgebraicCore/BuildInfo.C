//   Copyright (c)  2005,2007,2012  John Abbott

//   This file is part of the source of CoCoALib, the CoCoA Library.

//   CoCoALib is free software: you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation, either version 3 of the License, or
//   (at your option) any later version.

//   CoCoALib is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.

//   You should have received a copy of the GNU General Public License
//   along with CoCoALib.  If not, see <http://www.gnu.org/licenses/>.


#include "CoCoA/BuildInfo.H"
#include "CoCoA/config.H"

#include <limits>
#include <iostream>
using std::ostream;
using std::endl;
#include <string>
using std::string;

namespace CoCoA
{

  namespace BuildInfo
  {
    // Build info is actually passed in via three preprocessor variables:
    // COCOA_VERSION, COCOA_CXX, COCOA_CXXFLAGS -- all three are quoted strings.

    const std::string& version()
    {
      static const string info(COCOA_VERSION);
      return info;
    }

    const std::string& compiler()
    {
      static const string info(COCOA_CXX);
      return info;
    }

    const std::string& CompilationFlags()
    {
      static const string info(COCOA_CXXFLAGS);
      return info;
    }

    const std::string& CompilationDefines()
    {
      static const string info(COCOA_CXXFLAGS_DEFINES);
      return info;
    }


    void PrintAll(std::ostream& out)
    {
    // Below we use string literal juxtaposition, so that the build
    // info can readily be extracted from libcocoa.a by the command
    // strings libcocoa.a | egrep "CoCoA::BuildInfo"
      out << endl
          << "CoCoA::BuildInfo Summary of build information for CoCoALib:" << endl
          << "CoCoA::BuildInfo CoCoALib Version: " COCOA_VERSION << endl
          << "CoCoA::BuildInfo Compiler: " COCOA_CXX << endl
          << "CoCoA::BuildInfo Compilation Flags: " COCOA_CXXFLAGS << endl
          << "CoCoA::BuildInfo Compilation Flags Defines: " COCOA_CXXFLAGS_DEFINES << endl
          << "CoCoA::BuildInfo NumBits int = " << std::numeric_limits<unsigned int>::digits << endl
          << "CoCoA::BuildInfo NumBits long = " << std::numeric_limits<unsigned long>::digits << endl
          << "CoCoA::BuildInfo NumBits SmallExponent_t = " << std::numeric_limits<SmallExponent_t>::digits << endl
          << endl;
    }

  } // end of namespace CoCoA::BuildInfo

} // end of namespace CoCoA


// RCS header/log in the next few lines
// $Header: /Volumes/Home_1/cocoa/cvs-repository/CoCoALib-0.99/src/AlgebraicCore/BuildInfo.C,v 1.8 2012/10/24 12:10:45 abbott Exp $
// $Log: BuildInfo.C,v $
// Revision 1.8  2012/10/24 12:10:45  abbott
// Revised PrintAll: now prints also NumBits of SmallExponent_t
//
// Revision 1.7  2012/10/02 11:45:09  bigatti
// -- added CompilationDefines
//
// Revision 1.6  2012/10/02 10:36:12  abbott
// Revised interface to BuildInfo information strings.
// Several consequential changes.
//
// Revision 1.5  2012/06/20 12:26:26  bigatti
// -- now exporting also compiler and compilation flags
//
// Revision 1.4  2011/03/11 17:38:21  bigatti
// -- added missing include
//
// Revision 1.3  2011/03/11 17:18:55  bigatti
// -- added info on int and long size
//
// Revision 1.2  2007/10/30 17:14:08  abbott
// Changed licence from GPL-2 only to GPL-3 or later.
// New version for such an important change.
//
// Revision 1.1.1.1  2007/03/09 15:16:11  abbott
// Imported files
//
// Revision 1.4  2007/02/10 18:44:03  cocoa
// Added "const" twice to each test and example.
// Eliminated dependency on io.H in several files.
// Improved BuildInfo, and added an example about how to use it.
// Some other minor cleaning.
//
// Revision 1.3  2007/02/08 22:34:22  cocoa
// Changed BuildInfo: only BuildInfo version is publicly visible.
// Only BuildInfo needs complicated compile-time flags, so several changes
// to Makefiles etc.  Added a new example: ex-BuildInfo.C
//
// Revision 1.1.1.1  2005/10/17 10:46:54  cocoa
// Imported files
//
// Revision 1.2  2005/10/12 15:52:09  cocoa
// Completed test-RingFp1 and corrected/cleaned the SmallFp*
// and RingFp* files.
//
// Some minor tidying elsewhere.
//
// Revision 1.1  2005/10/06 16:36:42  cocoa
// Added the capability find out build information at run-time.
// The Makefiles should be a little tidier too.
//
