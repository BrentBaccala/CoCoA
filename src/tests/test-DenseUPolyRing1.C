#include "CoCoA/BuildInfo.H"
#include "CoCoA/DenseUPolyRing.H"
#include "CoCoA/FractionField.H"
#include "CoCoA/GlobalManager.H"
#include "CoCoA/QuotientRing.H"
#include "CoCoA/RingDenseUPolyClean.H"
#include "CoCoA/RingTwinFloat.H"
#include "CoCoA/RingQQ.H"
#include "CoCoA/RingZZ.H"
#include "CoCoA/VectorOperations.H"
#include "CoCoA/symbol.H"

using namespace CoCoA;

#include <iostream>
using std::cout;
using std::cerr;
using std::endl;
#include <vector>
using std::vector;

//----------------------------------------------------------------------
// Test for DenseUPolyRing functions on RingElem
// functions: +, -f, f-g, *, StdDeg, gcd, TidyGens, %, IsIrred
// environments: DenseUPolyClean (Fp, Z)
//----------------------------------------------------------------------

#define TEST_ASSERT(cond)   do { if (!(cond)) CoCoA::AssertionFailed(#cond, __FILE__, __LINE__); } while(0)


void TestDenseUPolyRing(DenseUPolyRing P)
{
  cout << "TESTING: " << P << endl << endl;

  if (!IsExact(P))
  {
    RingElem tf(NewRingTwinFloat(1024), 4);
    RingElem four(P);
    P->myRecvTwinFloat(raw(four), raw(tf));
    TEST_ASSERT(four == 4);
  }

  const RingElem& x = indet(P, 0);
  RingElem f1 = 3*x +23,
    f2 = IndetPower(P, 0, 4) + x,  // x^4 + x
    f3 = x + 29,
    f4(P);

  P->myAssignNonZeroCoeff(raw(f1), raw(one(CoeffRing(P))), 2); // x^2

  cout << "Given f1 = " << f1 << endl;
  cout << "  and f2 = " << f2 << endl << endl;
  cout << "f1+f2   gives  " << f1+f2 << endl;
  cout << "f2+f1   gives  " << f2+f1 << endl;
  cout << "f2-f1   gives  " << f2-f1 << endl;
  cout << "-f1   gives  "   << -f1 << endl;
  cout << "deg((2*x+1)*(3*x+1)) gives  "   << deg((2*x+1)*(3*x+1)) << endl;
  //   //  cout << "gcd(f1,f2)   gives  " << gcd(f1,f2) << endl;
  cout << "StdDeg(f1)   gives  " << StdDeg(f1) << endl;
  cout << "StdDeg(f2)   gives  " << StdDeg(f2) << endl;

  // if (IsTrueGCDDomain(P)) TEST_ASSERT(gcd(f1*f1f2, f2*f1f2) == f1f2);
  RingElem f1f2 = f1*f2;
  RingElem f2f1 = f2*f1;
  if (IsField(CoeffRing(P)) || IsTrueGCDDomain(CoeffRing(P)))
  {
    TEST_ASSERT( gcd(f1,f2) == 1 );
    TEST_ASSERT( IsInvertible(gcd(f1f2, f1*f3)/f1) );
    TEST_ASSERT( IsInvertible(gcd(f1f2, f2*f3)/f2) );
    TEST_ASSERT( IsInvertible(gcd(f1*f3, f2*f3)/f3) );
    TEST_ASSERT( IsInvertible(gcd(f1*f1f2, f2f1*f2)/(f1f2)) );
    TEST_ASSERT( IsDivisible(f1f2, f1) );
    TEST_ASSERT( IsDivisible(f1f2, f2) );
    TEST_ASSERT( !IsDivisible(f1, f2) );
    TEST_ASSERT( !IsDivisible(f2, f1) );
    TEST_ASSERT( (f1f2)/f1 == f2 );
    TEST_ASSERT( (f1f2)/f2 == f1 );
    TEST_ASSERT( deriv(f1,x) == 2*x+3 );
    try { std::cout << deriv(f1,x+1) << std::endl; }
    catch (const CoCoA::ErrorInfo& err) { if (err != ERR::NotIndet) throw; }
  }
  if (IsField(CoeffRing(P)))
  {
    try
    {
      TEST_ASSERT( IsIrred(f1) );
      TEST_ASSERT( !IsIrred(f2) );
      TEST_ASSERT( !IsIrred(f1f2) );
    }
    catch (const CoCoA::ErrorInfo& err) { if (err != ERR::NYI) throw; }
    RingElem g(P), acof(P), bcof(P);
    P->myExgcd(raw(g), raw(acof), raw(bcof), raw(f1), raw(f2));
    TEST_ASSERT(IsInvertible(g) || (g == gcd(f1,f2)));
    TEST_ASSERT(g == acof*f1 + bcof*f2);
  }
  
  P->myMulByXExp(raw(f1), 5);
  cout << "  P->myMulByXExp(raw(f1), 5) gives " << f1 << endl;
  P->myMulBy1MinusXExp(raw(f1), 2);
  cout << "  P->myMulBy1MinusXExp(raw(f1), 2) gives " << f1 << endl;
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2));
  cout << "  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives " << f1 << endl;
  
  cout << "one(P) = " << one(P) << endl;

  f1f2 = f1*f2;  // previous lines changed f1
  f2f1 = f2*f1;  // previous lines changed f1
  TEST_ASSERT(IsOne(one(P)));
  TEST_ASSERT(!IsOne(f1));
  TEST_ASSERT(!IsOne(f2));
  TEST_ASSERT(P->myIsValid(raw(f1)));
  TEST_ASSERT(P->myIsValid(raw(f2)));
  TEST_ASSERT(P->myIsValid(raw(f3)));
  TEST_ASSERT(P->myIsValid(raw(f4)));
  TEST_ASSERT(P->myIsValid(raw(f1f2)));
  TEST_ASSERT(P->myIsValid(raw(f2f1)));
  if (IsCommutative(P))  TEST_ASSERT(f1f2 == f2f1);
  TEST_ASSERT(f1 != f2);
  TEST_ASSERT(f1+f2 == f2+f1);
  TEST_ASSERT(f1+(-f1) == 0);
  TEST_ASSERT(f1-f1 == 0);
  TEST_ASSERT(f1f2 != f1);
  TEST_ASSERT(f1f2 != f2);
  
  if (IsField(CoeffRing(P)))
  {
    vector<RingElem> g;
    g.push_back(f1f2);
    g.push_back(f1*x);
    g.push_back(zero(P));
    g.push_back(f1*(x-1));
    ideal I = ideal(g);
    cout << "TidyGens(I) = " << TidyGens(I) << endl;
    TEST_ASSERT(!IsZero(I));
    TEST_ASSERT(IsZero(f1 % I));
    TEST_ASSERT(IsZero(f1f2 % I));
    TEST_ASSERT(!IsZero((f1f2+x) % I));
    
    ideal J1 = ideal(f1f2);
    ideal J2 = ideal(f1*(3*x-7));
    TEST_ASSERT(intersect(J1, J2) == ideal(f1f2*(3*x-7))); 
    TEST_ASSERT(colon(J1, J2) == ideal(f2)); 
 
    vector<RingElem> h(3, zero(P));
    TEST_ASSERT(TidyGens(ideal(h)).empty());
    TEST_ASSERT(IsZero(ideal(h)));
  }
  if ( IsFractionFieldOfGCDDomain(CoeffRing(P)) )
  {
    TEST_ASSERT( CommonDenom(x/3+one(P)/2) ==  6);
    TEST_ASSERT( ClearDenom(x/3+one(P)/2) ==  2*x+3);
  }
  TEST_ASSERT(IsZero(f4));
  f4 = 1;
  TEST_ASSERT(IsOne(f4));
  f2f1 = 1;
  TEST_ASSERT(P->myIsValid(raw(f2f1)));
  TEST_ASSERT(IsOne(f2f1));

  cout << "------------------------------------------------" << endl;
}


void program()
{
  GlobalManager CoCoAFoundations(UseNonNegResidues);

  const QuotientRing Fp = NewZZmod(101);
  const QuotientRing F6 = NewZZmod(6);
  const RingTwinFloat RR = NewRingTwinFloat(150);

  TestDenseUPolyRing(NewPolyRing_DUP(RingZZ()));
  TestDenseUPolyRing(NewPolyRing_DUP(Fp));
  TestDenseUPolyRing(NewPolyRing_DUP(F6, symbol("y")));
  TestDenseUPolyRing(NewPolyRing_DUP(RingQQ()));
  TestDenseUPolyRing(NewPolyRing_DUP(RR));
}


// Use main() to handle any uncaught exceptions and warn the user about them.
int main()
{
  try
  {
    program();
    return 0;
  }
  catch (const CoCoA::ErrorInfo& err)
  {
    cerr << "***ERROR***  UNCAUGHT CoCoA Error";
    ANNOUNCE(cerr, err);
  }
  catch (const std::exception& exc)
  {
    cerr << "***ERROR***  UNCAUGHT std::exception: " << exc.what() << endl;
  }
  catch(...)
  {
    cerr << "***ERROR***  UNCAUGHT UNKNOWN EXCEPTION" << endl;
  }

  BuildInfo::PrintAll(cerr);
  return 1;
}
