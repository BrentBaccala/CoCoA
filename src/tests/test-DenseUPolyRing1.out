TESTING: RingWithID(5,"ZZ[x]")

Given f1 = x^2 +3*x +23
  and f2 = x^4 +x

f1+f2   gives  x^4 +x^2 +4*x +23
f2+f1   gives  x^4 +x^2 +4*x +23
f2-f1   gives  x^4 -x^2 -2*x -23
-f1   gives  -x^2 -3*x -23
deg((2*x+1)*(3*x+1)) gives  2
StdDeg(f1)   gives  2
StdDeg(f2)   gives  4
  P->myMulByXExp(raw(f1), 5) gives x^7 +3*x^6 +23*x^5
  P->myMulBy1MinusXExp(raw(f1), 2) gives -x^9 -3*x^8 -22*x^7 +3*x^6 +23*x^5
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives -x^9 -3*x^8 -22*x^7 +4*x^6 +23*x^5 +x^3
one(P) = 1
------------------------------------------------
TESTING: RingWithID(11,"RingWithID(2)[x]")

Given f1 = x^2 +3*x +23
  and f2 = x^4 +x

f1+f2   gives  x^4 +x^2 +4*x +23
f2+f1   gives  x^4 +x^2 +4*x +23
f2-f1   gives  x^4 +100*x^2 +99*x +78
-f1   gives  100*x^2 +98*x +78
deg((2*x+1)*(3*x+1)) gives  2
StdDeg(f1)   gives  2
StdDeg(f2)   gives  4
  P->myMulByXExp(raw(f1), 5) gives x^7 +3*x^6 +23*x^5
  P->myMulBy1MinusXExp(raw(f1), 2) gives 100*x^9 +98*x^8 +79*x^7 +3*x^6 +23*x^5
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives 100*x^9 +98*x^8 +79*x^7 +4*x^6 +23*x^5 +x^3
one(P) = 1
TidyGens(I) = [100*x^9 +98*x^8 +79*x^7 +4*x^6 +23*x^5 +x^3]
------------------------------------------------
TESTING: RingWithID(15,"RingWithID(3)[y]")

Given f1 = y^2 +3*y +5
  and f2 = y^4 +y

f1+f2   gives  y^4 +y^2 +4*y +5
f2+f1   gives  y^4 +y^2 +4*y +5
f2-f1   gives  y^4 +5*y^2 +4*y +1
-f1   gives  5*y^2 +3*y +1
deg((2*x+1)*(3*x+1)) gives  1
StdDeg(f1)   gives  2
StdDeg(f2)   gives  4
  P->myMulByXExp(raw(f1), 5) gives y^7 +3*y^6 +5*y^5
  P->myMulBy1MinusXExp(raw(f1), 2) gives 5*y^9 +3*y^8 +2*y^7 +3*y^6 +5*y^5
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives 5*y^9 +3*y^8 +2*y^7 +4*y^6 +5*y^5 +y^3
one(P) = 1
------------------------------------------------
TESTING: RingWithID(16,"QQ[x]")

Given f1 = x^2 +3*x +23
  and f2 = x^4 +x

f1+f2   gives  x^4 +x^2 +4*x +23
f2+f1   gives  x^4 +x^2 +4*x +23
f2-f1   gives  x^4 -x^2 -2*x -23
-f1   gives  -x^2 -3*x -23
deg((2*x+1)*(3*x+1)) gives  2
StdDeg(f1)   gives  2
StdDeg(f2)   gives  4
  P->myMulByXExp(raw(f1), 5) gives x^7 +3*x^6 +23*x^5
  P->myMulBy1MinusXExp(raw(f1), 2) gives -x^9 -3*x^8 -22*x^7 +3*x^6 +23*x^5
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives -x^9 -3*x^8 -22*x^7 +4*x^6 +23*x^5 +x^3
one(P) = 1
TidyGens(I) = [-x^9 -3*x^8 -22*x^7 +4*x^6 +23*x^5 +x^3]
------------------------------------------------
TESTING: RingWithID(20,"RingWithID(4)[x]")

Given f1 = x^2 +3*x +23
  and f2 = x^4 +x

f1+f2   gives  x^4 +x^2 +4*x +23
f2+f1   gives  x^4 +x^2 +4*x +23
f2-f1   gives  x^4 -x^2 -2*x -23
-f1   gives  -x^2 -3*x -23
deg((2*x+1)*(3*x+1)) gives  2
StdDeg(f1)   gives  2
StdDeg(f2)   gives  4
  P->myMulByXExp(raw(f1), 5) gives x^7 +3*x^6 +23*x^5
  P->myMulBy1MinusXExp(raw(f1), 2) gives -x^9 -3*x^8 -22*x^7 +3*x^6 +23*x^5
  P->myAddMul(raw(f1), raw(LC(x)), 2, raw(f2)) gives -x^9 -3*x^8 -22*x^7 +4*x^6 +23*x^5 +x^3
one(P) = 1
TidyGens(I) = [-x^9 -3*x^8 -22*x^7 +4*x^6 +23*x^5 +x^3]
------------------------------------------------
